vpr/src/place/cut_spreader.cpp

```diff
@@ -0,0 +1,857 @@
+                            }
+                            if (input_len < best_inp_len) {
+                                best_inp_len = input_len;
+                                best_subtile = sub_t;
+                            }
+                            break;
+                        } else {
+                            if (bound != EMPTY_BLOCK_ID)
+                                remaining.emplace(imacro(bound) != -1 ? place_ctx.pl_macros[imacro(bound)].members.size() : 1, bound);
+                            ap->blk_locs[blk].loc = sub_t;
+                            placed = true;
+                            break;
+                        }
+                    }
+                }
+            } else {
+                for (auto sub_t : fast_t.at(nx).at(ny)) {
+                    std::vector<std::pair<ClusterBlockId, t_pl_loc>> targets;
+                    std::vector<std::pair<t_pl_loc, ClusterBlockId>> swaps_made;
+                    std::queue<std::pair<ClusterBlockId, t_pl_loc>> visit;
+                    visit.emplace(blk, sub_t);
+                    while (!visit.empty()) {
+                        ClusterBlockId vb = visit.front().first;
+                        VTR_ASSERT(!isPlaced(vb));
+                        t_pl_loc target = visit.front().second;
+                        visit.pop();
+                        auto log_type = clb_nlist.block_type(blk);
+                        auto result = std::find(log_type->equivalent_tiles.begin(), log_type->equivalent_tiles.end(), g_vpr_ctx.device().grid[target.x][target.y].type);
+                        if (result == log_type->equivalent_tiles.end()) goto fail;
+                        ClusterBlockId bound = place_ctx.grid_blocks[target.x][target.y].blocks[target.sub_tile];
+                        if (bound != EMPTY_BLOCK_ID)
+                            if (imacro(bound) != -1) goto fail;
+                        targets.emplace_back(vb, target);
+                        if (macro_head(vb) == vb) {
+                            const std::vector<t_pl_macro_member>& members = place_ctx.pl_macros[imacro(blk)].members;
+                            for (auto member = members.begin() + 1; member != members.end(); ++member) {
+                                t_pl_loc mloc = target + member->offset;
+                                visit.emplace(member->blk_index, mloc);
+                            }
+                        }
+                    }
+
+                    for (auto& target : targets) {
+                        ClusterBlockId bound = place_ctx.grid_blocks[target.second.x][target.second.y].blocks[target.second.sub_tile];
+                        if (bound != EMPTY_BLOCK_ID) unbindTile(target.second);
+                        bindTile(target.second, target.first);
+                        swaps_made.emplace_back(target.second, bound);
+                    }
+
+                    if (false) {
```

Comment:
My main use of goto is as a local exception. I'm not sure if a throw would get optimized to a jump. Maybe.

The other use is where I have something like:
```
if(...) {
  if(...) {
    goto doA;
    // otherwise this would be a copy of A
  } else {
    B
  }
} else {
doA:
   A
}
```

because I hate repeating `A`. Obviously this can be rewritten, but the rewritten condition gets harder to understand as the nesting gets more complex. The other option is to put A into a function and call that, but that can be annoying if A modifies a bunch of variables.

I've never seen jumping into `if (false) { ... }`, though. I was kind of surprised that would work.